include/proxy-wasm/context.h

```diff
@@ -0,0 +1,736 @@
+
+  /**
+   * Called on a Root Context to indicate that an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t token);
+
+  // Context
+
+  /**
+   * Call when a stream has completed (both sides have closed) or on a Root Context when the VM is
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone();
+
+  /**
+   * Call when a stream should log the final status of the stream. Occurs after onDone().
+   */
+  virtual void onLog();
+
+  /**
+   * Will be called on sever Wasm errors. Callees may report and handle the error (e.g. via an
+   * Exception) to prevent the proxy from crashing.
+   */
+  virtual void error(string_view message) {
+    std::cerr << message << "\n";
+    abort();
+  }
+
+  /**
+   * Called by all functions which are not overridden with a proxy-specific implementation.
+   */
+  virtual void unimplemented() { error("unimplemented proxy-wasm API"); }
+
+  /**
+   * Calls from the VM>
+   * This collection represents all the calls into the VM in the ABI
+   * (https://github.com/proxy-wasm/spec).  They are not broken into separate interfaces because
+   * doing so would result in duplication since calls into the VM must be overriden by the
+   * implementations in the ContextBase.
+   *
+   * Note: most of these return a WasmResult, assume:
+   * @return a WasmResult with the status of the call.
+   */
+
+  /**
+   * Log a message.  Note: log_prefix() is *not* auto-prepended end may be useful to provide some
+   * context.
+   */
+  virtual WasmResult log(uint64_t /* level */, string_view /* message */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Enables a periodic timer with the given period or sets the period of an existing timer. Note:
+   * the timer is associated with the Root Context of whatever Context this call was made on and
+   * there is only one timer available per Root Context.
+   * @param period is the period of the periodic timer in milliseconds.
+   */
+  virtual WasmResult setTimerPeriod(std::chrono::milliseconds /* period */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Provides the current time in nanoseconds.
+   */
+  virtual uint64_t getCurrentTimeNanoseconds() {
+    struct timespec tpe;
+    clock_gettime(CLOCK_REALTIME, &tpe);
+    uint64_t t = tpe.tv_sec;
+    t *= 1000000000;
+    t += tpe.tv_nsec;
+    return t;
+  }
+
+  /**
+   * Provides the status of the last call into the VM or out of the VM, similar to errno.
+   * @return the status code and a descriptive string.
```

Comment:
Is status code context specific? Why can't each call just return the status vs. the cludgy errno/getLastError() approach?

Reply:
They do, but they do not return the descriptive string.   This is a mechanism for providing more information, e.g. so that the plugin can send a that information to say some centralized monitoring system.